files/aws_lifecycle_hooks/tools.py

```diff
@@ -1,62 +1,104 @@
 #!/usr/bin/env python
-'''
-File managed by puppet in module aws_lifecycle_hooks
-'''
+"""File managed by puppet in module aws_lifecycle_hooks."""
 import json
 import functools
 import urllib.request
 import urllib.error
 import typing
 import yaml
 import attr
+import re
 
 from exceptions import ParsingError
 
 
 @attr.s
 class VolumeAttachment:
+    """Utility class."""
+
     volume_id = attr.ib()
     device_name = attr.ib()
 
 
+def metadata_version() -> str:
+    """AWS Metadata version string."""
+    return '2020-10-27'
+
+
+@functools.lru_cache(maxsize=1)
+def get_metadata_token() -> str:
+    """
+    Instance Metadata Service Version 2 (IMDSv2).
+
+    https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
+    Token fetching does not seem to work with versioned requests. Gives HTTP 403
+    """
+    headers = {'X-aws-ec2-metadata-token-ttl-seconds': 60}
+    req = urllib.request.Request(
+        url='http://169.254.169.254/latest/api/token',
+        headers=headers,
+        method='PUT',
+    )
+    with urllib.request.urlopen(req) as resp:
+        response_data = resp.read().decode("utf-8")
+    return response_data
+
+
+@functools.lru_cache(maxsize=1)
```

Comment:
you may want to enlarge this cache more than 1.
Right now, you only cache the last call.